### PR TITLE
docs: More detailed introduction for Option and Result

### PR DESCRIPTION
**Pull request by skarllot**
- More detailed introduction for Option and Result

### CODE REVIEW

File: docs/core/option.md [hunk 1]
```diff
@@ -1,7 +1,18 @@
 # Option
 
-The `Option` type is a replacement for nullable types. It can be used to represent a value that may or may not be
-present.
+The `Option` type encapsulates an optional value in a way that explicitly represents either the presence
+(represented as `Some`) or absence (represented as `None`) of a value. It can be used to replace nullable types
+and to deal with errors or exceptional cases without resorting to drastic measures such as throwing exceptions.
```
Comment (jeffijoe) [inline docs/core/option.md:5]: I’m not sure about emphasizing error _handling_ as a feature of Option, it’s really just a type-safe alternative to nullable types. Maybe describe how its simpler than for example dealing with potential nulls for reference types and default values for value types? Option lets you treat them the same, it doesn’t care whether its a reference or a value type.

What do you think?

Comment (skarllot) [reply docs/core/option.md:5]: Do you think that the second sentence should be dropped?

Haskell define Maybe as an error monad: https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Maybe.html

Comment (jeffijoe) [reply docs/core/option.md:5]: I see, the text looks very close to the Haskell one, but `Maybe` replaced with `Option`. If we're going to quote the Haskell docs, we should do so with a quote block and a link.

However, I don't agree with the Haskell docs on the point of it being for error handling.

How about the following intro text:

```md
The `Option` type is a powerful construct used to represent the presence or absence of a value. 
This type is particularly useful for eliminating null reference errors, a common source of bugs in C# applications. 

An `Option`  can be either `Some (containing a value), or `None` (indicating the absence of a value). 
By using `Option`, developers can enforce safer code practices, making it explicit when a variable, 
property, or parameter might not have a value and requires handling both cases.
```

**Comment by skarllot** [comment]
I've made some changes to the documentation

File: docs/core/result.md [hunk 2]
```diff
@@ -1,6 +1,26 @@
...
+a successful computation with a value or an error with a message or other error information.
+
+For optional values that do not carry errors, you should use the [`Option`](option) type.
+
+## Overview
+
+In many programming scenarios, operations can either succeed or fail. The `Result` type provides a type-safe
+way to handle these outcomes explicitly, avoiding throwing exceptions. Exceptions aren't are not good to
+report failures, because exception aren't part of methods signature and compiler doesn't help tracking
+unhandled exceptions.
```
Comment (jeffijoe) [inline docs/core/result.md:14]:
```suggestion
In many programming scenarios, operations can either succeed or fail. The `Result` type provides a type-safe
way to handle these outcomes explicitly, avoiding the use of exceptions. Exceptions are not ideal for
reporting failures, as exceptions are not part of a method's signature, and the compiler doesn't enforce
handling them. Additionally, exceptions are known to degrade performance.
```

**Review by jeffijoe — COMMENTED** [review]
Last bit, then we're good 👍

File: docs/core/result.md (hunk 2 above)
Comment (skarllot) [reply docs/core/result.md]: Done

**Comment by jeffijoe** [comment]
Thanks again @skarllot ! 🙏

**Pull request merged**